src/deserializer.rs

```diff
@@ -82,22 +82,24 @@ impl Deserializer {
     }
 
     fn _read_with_tag<T, F>(&mut self, expected_tag: u8, body: F) -> DeserializationResult<T>
-            where F: Fn(Vec<u8>) -> DeserializationResult<T> {
+            where F: Fn(&[u8]) -> DeserializationResult<T> {
         let tag = try!(self.reader.read_u8());
         // TODO: only some of the bits in the first byte are for the tag
         if tag != expected_tag {
             return Err(DeserializationError::UnexpectedTag);
         }
         let length = try!(self._read_length());
-        // Check for short data before allocating the data Vec so we disallow obscenely large
-        // lengths
-        if length > (self.reader.get_ref().len() - self.reader.position() as usize) {
-            return Err(DeserializationError::ShortData);
-        }
-        let mut data = vec![0; length];
-        let n = self.reader.read(&mut data).unwrap();
-        assert_eq!(n, length);
-        return body(data);
+
+        let result = {
+            let buf = self.reader.fill_buf().unwrap();
```

Comment:
I thought there would be a better way to do this. I first thought of [`Read::take`](https://doc.rust-lang.org/nightly/std/io/trait.Read.html#method.take), but that consumes the `reader`, which we don't want. Then I saw [`read_exact`](https://doc.rust-lang.org/nightly/std/io/trait.Read.html#method.read_exact), but that's unstable. So I guess this is okay for now.